contracts/allocation-vester/AllocationVester.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+// solhint-disable not-rely-on-time
+pragma solidity 0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import "../Claimable.sol";
+
+/**
+ * @title Allocation Vester
+ * @notice Allows allocating and distributing tokens to recipients by vesting
+ * it for arbitrary durations for each recipient
+ * @author shung for Pangolin
+ */
+contract AllocationVester is Claimable {
+    using SafeCast for uint256;
+    using SafeERC20 for IERC20;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct Member {
+        uint96 reserve; // remaining tokens to be distributed
+        uint96 stash; // tokens stashed when allocation is changed
+        uint64 lastUpdate; // timestamp of last harvest
+        uint256 rate; // rewards per second
+    }
+
+    /// @notice The mapping of member addresses to their properties
+    mapping(address => Member) public members;
+
+    /// @notice The set of members who have remaining allocation
+    EnumerableSet.AddressSet private _membersAddresses;
+
+    /// @notice The token to be distributed
+    IERC20 public immutable token;
+
+    /// @notice The total amount of tokens set to be streamed to all members
+    uint256 public reserve;
+
+    /// @notice The minimum duration a vesting can last
+    uint256 private constant MIN_DURATION = 8 weeks;
+
+    event AllocationSet(address indexed member, uint256 allocation, uint256 duration);
+    event Harvested(address indexed member, uint256 amount);
+    event Withdrawn(uint256 amount);
+
+    /**
+     * @notice Constructs a new AllocationVester contract
+     * @param distributionToken The address of the token to be distributed
+     * @dev Distribution token supply must fit 96bits
+     */
+    constructor(address distributionToken) {
+        require(distributionToken != address(0), "zero address");
+        token = IERC20(distributionToken);
+    }
+
+    /// @notice Claims all the pending rewards of the member
+    function harvest() external {
+        Member storage member = members[msg.sender];
+
+        // get the claimable rewards of the member
+        uint256 amount = pendingHarvest(msg.sender);
+        require(amount != 0, "no pending harvest");
+
+        // update the member's properties
+        member.lastUpdate = uint64(block.timestamp);
+        member.stash = 0;
+        member.reserve -= amount.toUint96();
```

Comment:
while `uint96` will accommodate substantially large numbers, if enough rewards are accumulated (especially for a token with larger than standard precision ie 18+ decimals) `harvest()` will cease to execute and funds cannot be claimed for that user

Reply:
answer from @Shungy:
> No rewards larger than max uint96 can be accumulated because reserves of the user can never be set larger than uint96 max by the admin. so there is no DOS risk on harvest. It would just be DOS when trying to setAllocations() with that large amounts